acceptancetests/assess_updateseries.py

```diff
@@ -0,0 +1,373 @@
+            # if 'systemd is not fully running, please reboot to start agents'\
+            #                 not in output:
+            #     raise JujuAssertionError("error running juju-updateseries "+\
+            #  "on {}: {}".format(unit, e))
+        else:
+            log.critical("test failure: reboot no longer before or after")
+            raise
+
+    log.info('sudo juju-updateseries ... --start-agents: {}'.format(output))
+
+    if reboot in 'after':
+        reboot_unit(client, unit, True)
+    else:
+        client.wait_for_started()
+    log.info("juju-updateseries .... --start-agents, reboot {} " +
+             "command, succeeded".format(reboot))
+
+
+def reboot_unit(client, unit, expect_agents):
+    """ Reboot a unit, wait for the agents to restart, or if not agents
+    restarting is not expected, wait for the ssh port to be available.
+
+    :param client: Juju client
+    :param unit: unit to reboot
+    :param expect_agents: bool, verify reboot via waiting for agents to
+    restart on unit.  If not, verify by waiting for port 22 being available.
+    :return: None
+    """
+    try:
+        log.info("Restarting unit: {}".format(unit))
+        cmd = build_ssh_cmd(client, unit, 'sudo shutdown -r now && exit')
+        output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        log.info("Restarting unit output: {}\n".format(output))
+    except subprocess.CalledProcessError as e:
+        # There is a failure logged with run with expect_agents False,
+        # though the reboot still happens???
+        logging.info(
+            "Error running shutdown:\nstdout: %s\nstderr: %s",
+            e.output, getattr(e, 'stderr', None))
+
+    if expect_agents:
+        log.info("wait_for_started()")
+        client.wait_for_started()
+    else:
+        # for cases where machine agents are not expected to be
+        # running after reboot, wait for the ssh port to be available
+        # instead.
+        status = client.get_status()
+        unit_status = dict(status.iter_units())
+        m = unit_status[unit]['machine']
+        machine_status = dict(status.iter_machines())
+        hostname = machine_status[m]['dns-name']
+        log.info("wait_for_port({}:22)".format(hostname))
+        wait_for_port(hostname, 22, timeout=240)
+
+
+def build_ssh_cmd(client, unit, command):
+    """ build_ssh_cmd is a helper method taking pieces from Client and Remote
+
+    :param client: Juju client
+    :param unit: unit to ssh to
+    :param command: command it be run via ssh
+    :return: completed command to which can be run
+    """
+    ssh_opts = [
+        "-o", "User ubuntu",
+        "-o", "UserKnownHostsFile /dev/null",
+        "-o", "StrictHostKeyChecking no",
+        "-o", "PasswordAuthentication no",
+    ]
+    cmd = ["ssh"]
+    cmd.extend(ssh_opts)
+    status = client.get_status()
+    unit_status = status.get_unit(unit)
+    cmd.extend([unit_status['public-address']])
+    cmd.extend([command])
```

Comment:
This could be simplified to, I guess the readability depends on how much go or python you've been looking at recently :-) (we're also not doing enough to worry about the performance issues of creating new lists instead of extending):
```python
cmd = ["ssh"] + ssh_opts + [unit_status['public-address']] + [command]
````

Reply:
done